### DISCUSSION TITLE
Browser shows gray screen although PDF file exists

### DISCUSSION THREAD

**Discussion by abid76** [cefsharp/CefSharp#4815] [General]
### Is there an existing issue for this?

- [X] I have searched both open/closed issues, no issue already exists.

### CefSharp Version

124.3.50

### Operating System

Windows 11

### Architecture

x64

### .Net Version

.Net 4.8

### Implementation

WPF

### Reproduction Steps

This happens only sometimes. I do not see any way to reproduce it.

### Expected behavior

Browser displays PDF file's content.

### Actual behavior

Browser shows a gray screen instead of file content.

![Screenshot 2024-05-21 164357](https://github.com/cefsharp/CefSharp/assets/34437604/46ee23b5-08a0-4b68-9cdf-99148a87de98)

E.g. this happened at 2024-05-21 16:37:32. But the cef-debug.log doesn't show any error except from some issue regarding an extension. I don't think this extension log message has anything to do with the issue described above as it is logged everytime you open a PDF file.

[cef-debug.log](https://github.com/cefsharp/CefSharp/files/15431731/cef-debug.log)

### Regression?

This also happend with version 123.0.60, but never with 106.0.260.

### Does this problem also occur in the CEF Sample Application

No.

### Other information

We're using CefSharp in a WPF `TabControl` to display PDF files which are located on a SMB share.

The files are processed using a `ResourceHandler` implementation (see below).

As this issue occurs only sometimes it may be some concurrency issue.

```
public override CefReturnValue ProcessRequestAsync(IRequest request, ICallback callback)
{
    string file = null;
    var uri = new Uri(request.Url);
    if (uri.Scheme == LocalSchemeHandlerFactory.SchemeName || uri.Scheme == SmbSchemeHandlerFactory.SchemeName)
    { 
        // smb://server/DMS/Dokumente/2020/03/2d18df48-04be-4b27-a141-dff5119debd3.docx.pdf
        var server = uri.Authority;
        var path = uri.LocalPath.Substring(1);
        file = Path.Combine(@"\\", server, path);
    }
    else
    {
        Log.Error($"Invalid Scheme: {uri.Scheme}");
    }

    if (file == null)
        return CefReturnValue.Cancel;

    Task.Run(() =>
    {
        using (callback)
        {
            if (!File.Exists(file))
            {
                callback.Cancel();
                return;
            }

            byte[] bytes = null;
            try
            {
                bytes = File.ReadAllBytes(file);
            }
            catch (Exception ex)
            {
                Log.Error($"Konnte Datei {file} nicht einlesen.", ex);
            }

            if (bytes == null)
            {
                Log.Debug("File is empty.");
                callback.Cancel();
            }
            else
            {
                var stream = new MemoryStream(bytes);
                stream.Position = 0;
                ResponseLength = stream.Length;

                var fileExtension = Path.GetExtension(file);
                MimeType = CefSharp.Cef.GetMimeType(fileExtension);
                StatusCode = (int)HttpStatusCode.OK;
                Stream = stream;

                callback.Continue();
            }
        }
    });

    return CefReturnValue.ContinueAsync;
}
```

**Comment by abid76** [comment]
My first question would be, what I can do to investigate this any further. Can it be a concurrency issue? Is there another way to process the file instead of starting a `Task` in `ProcessRequestAsync`?

**Reply by amaitland** [reply]
> My first question would be, what I can do to investigate this any further.

Have you checked the response in devtools? You can increase the log level. Also ask on https://magpcss.org/ceforum/index.php to see if any other CEF users are experiencing this.

The fact that the PDF viewer loaded suggests the request was completed.

> Can it be a concurrency issue?

From a cefsharp point of view there isn't any locking, if there was a concurrency issue it's likely in a lower layer.

> Is there another way to process the file instead of starting a Task in ProcessRequestAsync?

Yes, stream the date, read it in chunks. 

Return an error code to the browser rather than cancel the request. 

I'd move your logic out of the requesthandler

Have a look at https://github.com/cefsharp/CefSharp/blob/master/CefSharp/SchemeHandler/FolderSchemeHandlerFactory.cs#L90

**Comment by amaitland** [comment]
Converting to discussion for now, it's unlikely a cefsharp bug. A new issue can be opened if code changes are required.

**Comment by abid76** [comment]
> Have you checked the response in devtools? You can increase the log level. Also ask on https://magpcss.org/ceforum/index.php to see if any other CEF users are experiencing this.

As the error occured only to non-technical users the response hasn't been checked in devtools. I didn't find anything related in the forum.

I increased the log level to verbose. E.g. the error occured today at 11:37:35 am. You find the log file attached [cef-debug.log](https://github.com/user-attachments/files/15512707/cef-debug.log). 
There are quite a lot entries a this time, starting at `0531/113735.356`. 

There's also a warning logged (although this warning also appears when opening documents without any issues):
```
[0531/113735.966:WARNING:runtime_features.cc(730)] AttributionReportingCrossAppWeb cannot be enabled in this configuration. Use --enable-features=ConversionMeasurement,AttributionReportingCrossAppWeb in addition.
```

> Have a look at https://github.com/cefsharp/CefSharp/blob/master/CefSharp/SchemeHandler/FolderSchemeHandlerFactory.cs#L90

I implemented the request handling logic as suggested but the issue remains:
```csharp
public IResourceHandler Create(IBrowser browser, IFrame frame, string schemeName, IRequest request)
{
    if (schemeName != this.SchemeName)
    {
        string message = $"Invalid scheme '{schemeName}' in URL '{request.Url}'.";
        Log.Error(message);
        return ResourceHandler.ForErrorMessage(message, HttpStatusCode.BadRequest);
    }

    string file = ResolveFilePath(request);

    if (!File.Exists(file))
    {
        string message = $"File '{file}' doesn't exist.";
        Log.Warn(message);
        return ResourceHandler.ForErrorMessage(message, HttpStatusCode.NotFound);
    }

    Log.Debug("Open stream ...");
    var stream = new FileStream(file, FileMode.Open, FileAccess.Read, FileShare.Read);
    var extension = Path.GetExtension(file);
    var mimeType = CefSharp.Cef.GetMimeType(extension);
    Log.Debug("File processed. Returning resource handler.");
    return ResourceHandler.FromStream(stream, mimeType, autoDisposeStream: true);
}
``` 
The logs at the corresponding time show that the file is processed properly and the last statement `return ResourceHandler.FromStream(stream, mimeType, autoDisposeStream: true);` is executed.

There maybe an issue loading the file from the SMB share. But then I would expect that `File.Exists` returns `false` or at least some error is shown in the logs.

**EDIT:** I wonder how to find the root cause of this issue. Is there a way to open devtools when the issue shows up (although this maybe too late)? Is there another way to monitor the process when CEF loads and displays the file?

**Reply by amaitland** [reply]
> There maybe an issue loading the file from the SMB share. But then I would expect that File.Exists returns false or at least some error is shown in the logs.

Personally I'd be trying to cache files locally

> Also ask on https://magpcss.org/ceforum/index.php to see if any other CEF users are experiencing this.

Try asking on ceforum for guidance.

**Comment by abid76** [comment]
Thanks for your help, I'll give the local caching a try.

> Try asking on ceforum for guidance.

My account is disabled. It tells me to contact a board administrator. But I didn't find a contact form or anything. It seems, without being logged in it's not possible to get in touch with an admin. My username is sugar76. May you have an idea what to do?

**Reply by amaitland** [reply]
https://github.com/chromiumembedded/cef/issues/2962#issuecomment-1465023905

**Reply by abid76** [reply]
> Personally I'd be trying to cache files locally

Unfortunately the issue remains when caching the files locally. I'm using this code
```csharp
bool delete = false;
Stream stream;
if (SmbSchemeHandlerFactory.SchemeNameSmb == schemeName) // in most cases the file comes from a SMB share
{
    file = FileService.CopyToTempDir(file); // FileService is an application specific class which uses File.Copy
    stream = new MemoryStream(File.ReadAllBytes(file));
    delete = true;
}
else
{
    stream = new FileStream(file, FileMode.Open, FileAccess.Read, FileShare.Read);
}

var extension = Path.GetExtension(file);
var mimeType = CefSharp.Cef.GetMimeType(extension);
try
{
    return ResourceHandler.FromStream(stream, mimeType, autoDisposeStream: true);
}
finally
{
    if (delete)
    {
        FileService.Delete(file);
    }
}
```

> Try asking on ceforum for guidance.

I asked in the forum (waiting for the post to be approved)-

**Reply by amaitland** [reply]
Asked at https://www.magpcss.org/ceforum/viewtopic.php?f=6&t=19819

Continue the discussion there.
